lib/generics.go

```diff
@@ -0,0 +1,60 @@
+package lib
+
+import "sort"
+
+// Generic Set object. Retains the order elements are addd to the set.
+// Use set.ToOrderedSlice() to retrieve in original order.
+type Set[T comparable] struct {
+	_innerMap map[T]int
+}
+
+func NewSet[T comparable](elements []T) *Set[T] {
+	set := &Set[T]{_innerMap: make(map[T]int)}
+	for _, element := range elements {
+		set.Add(element)
+	}
+	return set
+}
+
+func (set *Set[T]) Add(element T) {
+	// Add element to set.
+	if !set.Includes(element) {
+		set._innerMap[element] = set.Size()
```

Comment:
hmm isn't it possible that we'll have multiple keys with the same value if we have adds and removes? Maybe a better way to do this to ensure ordering would be to have a monotonically increasing integer in the set struct and just add 1 to it with each add operation (and never decrease it with deletes).

Reply:
Yeah, that's a good idea. Will add that.